Weirdly enough, this extension is not finding its own Intelephense installation

Hi @mikebronner 👋

Recently I noticed that your Intelephense extension stopped working (I cannot pinpoint exactly _when_ that first happened, but my guess is that it stopped working around the time Nova upgraded to 8; I'm currently running Nova 8.3 [Build 343365] under macOS Big Sur 11.6.3 [20G405]).

These are the logs I get (restarting the extension will basically print out the same errors, just on a different workspace storage path). Note that the red exclamation mark is the closest I could come with a symbol that marks the error in Nova (the white exclamation mark inside a red circle).

```
   Intelephense[20:14:43.906000] Creating extension workspace storage path: /Users/gwyneth/Library/Application Support/Nova/Workspaces/5901757A-1876-40A2-A83F-89948CB68871/genealabs.intelephense
   Intelephense[20:14:43.906000] Creating extension global storage path: /Users/gwyneth/Library/Application Support/Nova/Extensions/genealabs.intelephense
   Intelephense[20:14:43.906000] Creating Intelephense workspace storage path: /Users/gwyneth/Library/Application Support/Nova/Workspaces/5901757A-1876-40A2-A83F-89948CB68871/genealabs.intelephense/intelephense
   Intelephense[20:14:43.906000] Creating Intelephense global storage path: /Users/gwyneth/Library/Application Support/Nova/Extensions/genealabs.intelephense/intelephense
   Intelephense[20:14:44.899000] Starting Intelephense installation.
❗️ Intelephense LSP[20:14:48.626000] node:internal/modules/cjs/loader:936
❗️ Intelephense LSP[20:14:48.626000]   throw err;
❗️ Intelephense LSP[20:14:48.626000]   ^
❗️ Intelephense LSP[20:14:48.626000] 
❗️ Intelephense LSP[20:14:48.626000] Error: Cannot find module '/Users/gwyneth/Library/Application Support/Nova/Workspaces/5901757A-1876-40A2-A83F-89948CB68871/genealabs.intelephense/node_modules/intelephense/lib/intelephense.js'
❗️ Intelephense LSP[20:14:48.627000]     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
❗️ Intelephense LSP[20:14:48.627000]     at Function.Module._load (node:internal/modules/cjs/loader:778:27)
❗️ Intelephense LSP[20:14:48.627000]     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
❗️ Intelephense LSP[20:14:48.627000]     at node:internal/main/run_main_module:17:47 {
❗️ Intelephense LSP[20:14:48.627000]   code: 'MODULE_NOT_FOUND',
❗️ Intelephense LSP[20:14:48.627000]   requireStack: []
❗️ Intelephense LSP[20:14:48.627000] }
❗️ Intelephense LSP[20:14:48.627000] 
❗️ Intelephense LSP[20:14:48.627000] Node.js v17.3.0
```

I have looked through your code just very superficially, so I might be completely out of my league here, but I believe that, for some reason, `npm` is failing to create the required `node_modules` directory inside the workspace.

Indeed, the directory `/Users/gwyneth/Library/Application Support/Nova/Workspaces/5901757A-1876-40A2-A83F-89948CB68871/genealabs.intelephense/intelephense` _does_ exist — i.e. it _was_ properly created, with 755 permissions (which is about right) and the correct user owning it.

Now, I have several possible explanations for this behaviour, but sadly I cannot figure out to fix them, even if my hypothesis were correct...

1. I'm currently running PHP 8.1. While this ought not to interfere with `npm`, your README explicitly mentions that Intelephense does not yet support PHP 8+. For my purposes, it's pretty much irrelevant; while I use some very minor things that appeared in 7.3 or 7.4, I _mostly_ stick to 'old school' PHP, so I wouldn't really worry much about the lack of support for 8+. And the errors on the Nova log do not seem to point to _that_, anyway...
2. Almost all CLI software I've got on my Mac (and some of the GUI-based as well) has been installed via Homebrew. In particular, `nodejs` and `npm` — because I'm not overly familiar with hard-core JavaScript coding — have been installed, but for many months (years?) I have long suspected that there is something wrong with the configuration. Ironically, I only noticed that when installing one Nova extension, which looked for the installed `npm` binary in the _right_ place, but... it wasn't there! (Every time  either`nodejs` or `npm` get upgraded, I need to remember to add a symbolic link for `npm`). This _could_ have some consequences for your code, which may be expecting a rational layout for `nodejs`, and my Mac just happens to have a very messy installation (but, as said, I don't know how to fix it!).

Also, I wonder if there is a way to use the system-wide Intelephense installation, instead of getting Nova to install its own...

Anyway, any insights into this would be most appreciated! Thanks so much in advance!

--- a/.gitignore
+++ b/.gitignore
@@ -1,1 +1,2 @@
 .DS_Store
+.nova

--- a/.nova/Configuration.json
+++ b/.nova/Configuration.json
@@ -1,3 +1,5 @@
 {
-  "workspace.color" : 1
+  "editor.default_syntax" : "javascript",
+  "workspace.art_style" : 1,
+  "workspace.color" : 8
 }

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,14 +2,24 @@
 ## [Upcoming]
 - Add language server configuration settings.
 
+## [0.1.8] - 2022-04-22
+### Changed
+- During the first installation, a `package.json` file will be generated from scratch, in
+  order to allow `npm` to do an installation in the correct place.
+### Fixed
+- Some path names were inconsistent.
+### Updated
+- [README.md](README.md) includes some information about the caveats of the automatic
+  installation process for Inteliphense using `npm`.
+
 ## [0.1.7] - 2022-04-21
 ### Removed
 - outdated Known Issues from README.
 
 ## [0.1.6] - 2022-04-21
-  ### Updated
-  - intelephense version to 1.8.2. This resolved the initialization errors and now
-    also provides PHP 8 support.
+### Updated
+- intelephense version to 1.8.2. This resolved the initialization errors and now
+  also provides PHP 8 support.
 
 ## [0.1.5] - 2021-04-26
 ### Reverted

--- a/LICENSE.md
+++ b/LICENSE.md
@@ -1,6 +1,6 @@
 # The MIT License (MIT)
 
-Copyright © 2021 GeneaLabs, Mike Bronner.
+Copyright © 2021,2022 GeneaLabs, Mike Bronner.
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 

--- a/README.md
+++ b/README.md
@@ -77,7 +77,7 @@ During activation, this extension will attempt to deploy on the workspace a full
 
 In some circumstances (under investigation!), `npm` may fail to create the directories it needs (notably, `node_modules`) in the place where Nova (and, later, Intelephense) expects them; as a consequence, the LSP will fail with a console error that might not be easy to debug — since it depends quite a lot on how you are currently managing `npm`, both at the global as well as the local level.
 
-This step may require some improvement in the future; alternatively, there might be a choice to use the _global_ installation of `npm` instead, or, if you have one local `npm` installation in your current PHP project, this extension might attempt to use it. Such options will require further options on the Preferences panel; stay tuned...
+This step may require some improvement in the future; alternatively, there might be a choice to use the _global_ installation of `npm` instead, or, if you have one local `npm` installation in your current PHP project, this extension might attempt to use it. Such options will require a few new options on the Preferences panel; stay tuned...
 
 <!--
 👋 That's it! Happy developing!

--- a/README.md
+++ b/README.md
@@ -71,6 +71,14 @@ To configure global preferences, open **Extensions → Extension Library...** th
 
 You can also configure preferences on a per-project basis in **Project → Project Settings...**
 
+## Caveats
+
+During activation, this extension will attempt to deploy on the workspace a fully functional (and independently managed) installation of Intelephense; you will need to have `npm` installed (you can get it with `brew install npm`).
+
+In some circumstances (under investigation!), `npm` may fail to create the directories it needs (notably, `node_modules`) in the place where Nova (and, later, Intelephense) expects them; as a consequence, the LSP will fail with a console error that might not be easy to debug — since it depends quite a lot on how you are currently managing `npm`, both at the global as well as the local level.
+
+This step may require some improvement in the future; alternatively, there might be a choice to use the _global_ installation of `npm` instead, or, if you have one local `npm` installation in your current PHP project, this extension might attempt to use it. Such options will require further options on the Preferences panel; stay tuned...
+
 <!--
 👋 That's it! Happy developing!
 

--- a/Scripts/main.js
+++ b/Scripts/main.js
@@ -1,23 +1,28 @@
 var workspaceStoragePath = nova.path.join(nova.extension.workspaceStoragePath, "intelephense");
 var globalStoragePath = nova.path.join(nova.extension.globalStoragePath, "intelephense");
 var langserver = null;
-var requiredVersion = "1.8.2";
+var requiredVersion = "1.8.2";  // for Inteliphense
+var packageJSONfile = nova.path.join(workspaceStoragePath, "package.json"); // Create it if it doesn't exist, or npm will silently fail (gwyneth 20220422)
+
 var install = async function () {
+    if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
+        console.log("install: Launching 'npm' in directory:", workspaceStoragePath);
+    }
     installation = new Process(
         "/usr/bin/env",
         {
             args: ["npm", "install", "intelephense@" + requiredVersion],
-            cwd: nova.extension.workspaceStoragePath,
+            cwd: /* nova.extension.workspaceStoragePath */ workspaceStoragePath,
         }
     );
     installation.onStdout(function (output) {
         if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
-            console.log("Installation Output", output.trim());
+            console.log("Installation Output:", output.trim());
         }
     });
     installation.onDidExit(function (status) {
         if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
-            console.log("Installation completed with status", status);
+            console.log("Installation completed with status:", status);
         }
 
         langserver = new IntelephenseLanguageServer();
@@ -30,19 +35,61 @@ var install = async function () {
     installation.start();
 };
 var installIfMissing = async function () {
+    /* check first if we have a valid `package.json` file in the right place */
+    if (nova.fs.access(packageJSONfile, nova.fs.F_OK) === false) {
+        if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
+            console.log("installIfMissing: 'package.json' not found, creating a default one.");
+        }
+        /*  Now create our own `package.json` from scratch, using the defaults below.
+         *  Note: I'm pretty sure there must be a more idiomatic JS way to do this;
+         *  also: nothing is done concurrently, since this whole function is already async anyway.
+         *  (gwyneth 20220422)
+         */
+        fileContent = `{
+          "name": "` + nova.extension.name + `",
+          "description": "` + nova.extension.description + `",
+          "version": "` + nova.extension.version + `",
+          "repository": {
+            "type": "git",
+            "url": "https://github.com/GeneaLabs/panic-nova-intelephense"
+          },
+          "keywords": ["languages"],
+          "author": "` + nova.extension.vendor + `",
+          "license": "MIT",
+          "bugs": {
+            "url": "https://github.com/GeneaLabs/panic-nova-intelephense/issues"
+          },
+          "homepage": "https://extensions.panic.com/extensions/genealabs/genealabs.intelephense/"
+        }`;
+        try {
+            let fileJSON = nova.fs.open(packageJSONfile, 'wt');
+            fileJSON.write(fileContent);
+            fileJSON.close();
+        } catch(error) {
+            console.warn("installIfMissing: could not create 'package.json' file, error was: ", error);
+        }
+    } else  {
+        if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
+            console.log("installIfMissing: readable 'package.json' found, continuing install with 'npm'...");
+        }
+    }
+
     let installedVersion = "";
+    if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
+        console.log("installIfMissing: Launching npm in path:", workspaceStoragePath);
+    }
     let process = new Process(
         "/usr/bin/env",
         {
             args: ["npm", "list", "--depth 0", "--parseable", "--long", "intelephense"],
-            cwd: nova.extension.workspaceStoragePath,
+            cwd: /* nova.extension.workspaceStoragePath */ workspaceStoragePath,
         }
     );
     process.onStdout(function (output) {
         installedVersion = ((output.split(":")[1] || "").split("@")[1] || "");
 
         if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
-            console.log("Currenlty installed version", installedVersion);
+            console.log("installIfMissing: Intelephense currently installed version:", installedVersion);
         }
     })
     process.onDidExit(function (status) {
@@ -51,7 +98,9 @@ var installIfMissing = async function () {
 
             return;
         }
-
+        if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
+            console.log("installIfMissing: Checking installation did complete with status:", status);
+        }
         langserver = new IntelephenseLanguageServer();
     });
     process.start();
@@ -74,7 +123,7 @@ var activate = async function () {
         nova.fs.mkdir(workspaceStoragePath);
         nova.fs.mkdir(globalStoragePath);
     } catch (error) {
-        console.error("Directory creation failed:", error);
+        console.error("Directory creation failed during activation:", error);
     }
 
     await installIfMissing();
@@ -109,6 +158,12 @@ class IntelephenseLanguageServer {
         this.start();
     }
 
+    activate() {
+        if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
+            console.log("Activating language server using:", workspaceStoragePath);
+        }
+    }
+
     deactivate() {
         if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
             console.log("Deactivating language server.");
@@ -124,12 +179,12 @@ class IntelephenseLanguageServer {
         }
 
         if (nova.config.get('genealabs.intelephense.debugging', 'boolean')) {
-            console.log("Starting language server.");
+            console.log("Starting language server with path:", workspaceStoragePath);
         }
 
         if (!path) {
             path = nova.path.join(
-                nova.extension.workspaceStoragePath,
+                workspaceStoragePath,
                 'node_modules',
                 'intelephense',
                 'lib',

--- a/extension.json
+++ b/extension.json
@@ -3,7 +3,7 @@
     "name": "Intelephense",
     "organization": "GeneaLabs",
     "description": "Nova wrapper of the fabulous Intelephense language server from Ben Mewburn.",
-    "version": "0.1.7",
+    "version": "0.1.8",
     "categories": ["languages"],
 
     "main": "main.js",

